subcommittee/coding-guidelines/initiatives/safe-use-of-unsafe-guidelines/rust-embedded-lld-safe-definition.md

```diff
@@ -0,0 +1,295 @@
+and for clock system of microcontroller due to sensitivity to physical
+conditions of device and surrounding environments (e.g., variance of
+process manufacturing, operating temperature…). Therefore, undefined
+behavior is the worst-case scenario and there is no software test that
+can detect it.
+
+It shall be highlighted that an undefined behavior could be triggered
+not only by wrong value of registers but also by wrong order of reading
+and writing of registers.
+
+## Rust safe/unsafe are not well defined when hardware is involved.
+
+From previous discussion it could appear obvious that safe low level
+drivers shall protect from any misuses that could trigger undefined
+behavior even from hardware.
+
+Unfortunately, this is not generally accepted and there are different
+understandings. Let`s list them:
+
+### Svd2rust
+
+Svd2Rust[^6] generated peripheral access crates (PAC) from xml register
+description. Peripheral access crates provides API to write and read
+peripheral registers.
+
+Safe functions provided by generated crates protect users from following
+issue:
+
+  - Writing out of range value for bitfields.
+
+  - Concurrent access to same register from different interrupt/thread
+
+  - Read or write registers that cannot be read or written.
+
+_These functions cannot protect from undefined
+behavior triggered by improper order of calling functions_.
+
+e.g. Clock initialization, peripheral initialization.
+
+This approach seems conflicting with other approach:
+
+  - In Embedonomicon is described how to create safe API for DMA low
+    level driver[^7]. Why shall I take all these measures if all my
+    register access API are safe?
+
+  - All C library all functions are by considered by default unsafe. Why
+    a very low level access to external HW function is considered safer
+    than accessing a C library.
+
+On svd2rust and safety there is an ongoing discussion in an issue of
+GitHub project[^8] related to this topic.
+
+### Is safe and unsafe not applicable to external world?
+
+I have talked with some developers, and they don`t consider relevant HW
+undefined behavior because are not part of Rust developed software. In
+other words, they assume that unsafety shall be restricted to only rust
+code memory issues, thread issue but not to HW component.
+
+This position is similar to the one regarding the safety of *fs::open*
+in Unix system[^9]. The *fs::open* is considered safe function even if
+it can open special file */proc/self/mem* that allows unchecked full
+access to process memory. Moreover, it is stated the following in the
+documentation:
+
+*Rust`s safety guarantees only cover what the program itself can do, and
+not what entities outside the program can do to it.*
+
+However, I still see the same conflicting usage of unsafe as in the
+previous chapter:
+
+  - C API are by default unsafe. Are not referring to external C
+    program? If the previous definition holds, they should be considered
+    safe.
```

Comment:
My thought here would be that because as you defined for the unsafe block,

> the contract necessary to call the operations inside the block has been checked by the programmer and is guaranteed to be respected

Any calls made via C API should be considered default unsafe until a programmer checks to ensure the contract of the API is fulfilled and Rust's safety guarantees will be met.

Reply:
Sorry but I don't fully understand this comment.
I am trying to explain that there is a contradiction between the statement:

> *Rust`s safety guarantees only cover what the program itself can do, and
not what entities outside the program can do to it.*

and the fact that FFI (external word) *are* considered unsafe by the compiler. 
Moreover the definition of unsafe block is copied from [here](https://doc.rust-lang.org/std/keyword.unsafe.html#the-different-meanings-of-unsafe)

Are you proposing a change in the text ?

Reply:
I was not proposing a change in the text. I just didn't understand that you were setting up a contradiction there.

I understand what you are saying now, after reading your reply. Perhaps you could consider making that section more explicit, or you could just leave it as is.